src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -46,80 +46,19 @@ internal GeneratorDriver(ParseOptions parseOptions, ImmutableArray<ISourceGenera
         }
 
-        public GeneratorDriver RunFullGeneration(Compilation compilation, out Compilation outputCompilation, out ImmutableArray<Diagnostic> diagnostics, CancellationToken cancellationToken = default)
+        public GeneratorDriver RunGenerators(Compilation compilation, CancellationToken cancellationToken = default)
         {
-            // with no generators, there is no work to do
-            if (_state.Generators.Length == 0)
-            {
-                outputCompilation = compilation;
-                diagnostics = ImmutableArray<Diagnostic>.Empty;
-                return this;
-            }
-
-            // run the actual generation
-            var state = StateWithPendingEditsApplied(_state);
-            var stateBuilder = ArrayBuilder<GeneratorState>.GetInstance();
-            var receivers = PooledDictionary<ISourceGenerator, ISyntaxReceiver>.GetInstance();
-            var diagnosticsBag = new DiagnosticBag();
-
-            for (int i = 0; i < state.Generators.Length; i++)
-            {
-                var generator = state.Generators[i];
-                var generatorState = state.GeneratorStates[i];
-
-                // initialize the generator if needed
-                if (!generatorState.Info.Initialized)
-                {
-                    generatorState = InitializeGenerator(generator, diagnosticsBag, cancellationToken);
-                }
-                stateBuilder.Add(generatorState);
-
-                // create the syntax receiver if requested
-                if (generatorState.Info.SyntaxReceiverCreator is object)
-                {
-                    var rx = generatorState.Info.SyntaxReceiverCreator();
-                    receivers.Add(generator, rx);
-                }
-            }
-
-            // Run a syntax walk if any of the generators requested it
-            if (receivers.Count > 0)
-            {
-                GeneratorSyntaxWalker walker = new GeneratorSyntaxWalker(receivers.Values.ToImmutableArray());
-                foreach (var syntaxTree in compilation.SyntaxTrees)
-                {
-                    walker.Visit(syntaxTree.GetRoot(cancellationToken));
-                }
-            }
-
-            // https://github.com/dotnet/roslyn/issues/42629: should be possible to parallelize this
-            for (int i = 0; i < state.Generators.Length; i++)
-            {
-                var generator = state.Generators[i];
-                var generatorState = stateBuilder[i];
-                try
-                {
-                    // don't try and generate if initialization failed
-                    if (!generatorState.Info.Initialized)
-                    {
-                        continue;
-                    }
+            var state = RunGeneratorsCore(compilation, new DiagnosticBag(), cancellationToken);
```

Comment:
>new DiagnosticBag() [](start = 55, length = 19)

Where are these diagnostics recorded? Also, should we pass in a pooled instance here?